### PR TITLE
fix(client): Only send postMessages to and from the expected parent when using the iframe.

### LINKED ISSUE #2151: iframe flow: check client_id's expected origin before `ping` postMessage
Only send the postMessage to the expected origin. If no response, fail, we are illegally iframed.

### PR DESCRIPTION
**Pull request by shane-tomlinson**
If the content server is illegally iframed, no content will be displayed to the user until the channel times out, and then a 500 error will be displayed. There is no longer a special "illegal iframe" screen, so all of that logic is removed.

fixes #2151

### CODE REVIEW

**Comment by shane-tomlinson** [comment]
Ref #2146

File: app/scripts/lib/channels/mixins/postmessage_receiver.js [hunk 1]
```diff
@@ -47,6 +47,8 @@ define([
 
       this.window = options.window || window;
 
+      this._origin = options.origin || '*';
```
Comment (shane-tomlinson) [inline app/scripts/lib/channels/mixins/postmessage_receiver.js]: `*` is used by the fx-desktop channel. Maybe it should set it manually instead of setting it to `*` by default.

File: app/scripts/lib/channels/mixins/postmessage_receiver.js [hunk 2]
```diff
@@ -90,9 +92,20 @@ define([
       errorIfNoResponse.call(this, outstanding);
     },
 
+    isEventFromValidOrigin: function (event) {
+      if ((! this._origin) ||
```
Comment (shane-tomlinson) [inline app/scripts/lib/channels/mixins/postmessage_receiver.js]: Is there ever not an origin set?

File: app/scripts/models/auth_brokers/iframe.js [hunk 3]
```diff
@@ -28,27 +28,38 @@ define([
...
+    return getExpectedParentOrigin(self.relier)
+      .then(function (expectedOrigin) {
+        self.getChannel().communicateWithOrigin(expectedOrigin);
+
+        // No message will be received if the parent is either:
+        // 1. not set up to respond correctly
+        // 2. not the expected origin
+        //
+        // In either case, show a 500 error
+        return getActualParentOrigin(self);
```
Comment (shane-tomlinson) [inline app/scripts/models/auth_brokers/iframe.js]: This should be renamed to `ensureExpectedOriginResponds` or similar.

**Comment by ncalexan** [comment]
@shane-tomlinson asked privately what the `postMessage` origin used by the firefox-ios client to respond to messages like `can_link_account` currently is.  We feed back whatever URL we connected to, pretending that we are the page answering itself.  (I assumed the page would not accept "*".)

The code is at [1]; `self.url` in this context is like [2].

[1] https://github.com/mozilla/firefox-ios/blob/66c8d746f0c8ca0f3c03122663205276b8ed85ab/FxAClient/Frontend/SignIn/FxAContentViewController.swift#L192
[2] https://github.com/mozilla/firefox-ios/blob/66c8d746f0c8ca0f3c03122663205276b8ed85ab/Providers/FirefoxAccountConfiguration.swift#L26

**Comment by shane-tomlinson** [comment]
Thanks for the info @ncalexan. I'm trying to figure out how to best handle this.

File: app/scripts/models/auth_brokers/fx-desktop.js [hunk 4]
```diff
@@ -93,7 +93,10 @@ define([
       var channel = this._channel || new FxDesktopChannel();
 
       channel.init({
-        window: this.window
+        window: this.window,
+        // The browser is the origin. Events from the browser have no origin.
```
Comment (shane-tomlinson) [inline app/scripts/models/auth_brokers/fx-desktop.js]: postMessages that originate in Firefox Desktop have an origin of the string `null`. @ncalexan points out that Fx for iOS has an origin of the page itself. It would be nice to have this handled in an abstracted manner.

Relying on the string `null` seems brittle to changes in the Gecko code. @ckarlof, @vladikoff and @rfk - suggestions?

Note, `*` is the same as the current behavior.

File: app/scripts/lib/channels/mixins/postmessage_receiver.js [hunk 5]
```diff
@@ -90,9 +92,19 @@ define([
...
+      }
+
+      return this._origin === event.origin;
+    },
+
     receiveMessage: function (event) {
-      // TODO - Ensure the event.origin is one we trust,
-      // esp in the iframe case
+      if (! this.isEventFromExpectedOrigin(event)) {
+        // from an unexpected origin, drop it on the ground.
```
Comment (shane-tomlinson) [inline app/scripts/lib/channels/mixins/postmessage_receiver.js]: Should messages from unexpected origins be logged to metrics?

File: app/scripts/models/auth_brokers/fx-desktop.js (hunk 4 above)
Comment (shane-tomlinson) [inline app/scripts/models/auth_brokers/fx-desktop.js]: For more context - communication with Firefox Desktop for Sync sends [custom events](https://github.com/mozilla/fxa-content-server/blob/c46d67333bd72b74ad3f488c57952635475c7c13/app/scripts/lib/channels/fx-desktop.js#L17) to Firefox, Firefox responds by using [postMessage](https://dxr.mozilla.org/mozilla-central/source/browser/base/content/aboutaccounts/aboutaccounts.js#256). 

The messages that are sent by Firefox have an origin of `null`. This is [known behavior](https://bugzilla.mozilla.org/show_bug.cgi?id=1040257). A quick check shows that at least in Fx 37, message events from Firefox have a field `isTrusted: true`. message events from iframe reliers have `isTrusted: false`. Perhaps this is a viable alternative check.

I will check to see if Fx 29 (the first version of Fx desktop with new Sync) has the same behavior.

Comment (shane-tomlinson) [inline app/scripts/models/auth_brokers/fx-desktop.js]: More info on [isTrusted](https://developer.mozilla.org/docs/Web/API/Event/isTrusted).

File: app/scripts/lib/channels/mixins/postmessage_receiver.js (hunk 5 above)
Comment (rfk) [inline app/scripts/lib/channels/mixins/postmessage_receiver.js]: yep, this seems sensible

File: app/scripts/models/auth_brokers/fx-desktop.js (hunk 4 above)
Comment (rfk) [inline app/scripts/models/auth_brokers/fx-desktop.js]: Hmm, it sound like `isTrusted` isn't *quite* right then.

ISTM that the string value `"null"`, while fragile, will at least fail securely - if we happen to see it, then we're sure the event came from trusted desktop code, right?  So a "null OR expected-origin" check, although ugly, would be safe.

File: app/scripts/lib/channels/mixins/postmessage_receiver.js (hunk 5 above)
Comment (shane-tomlinson) [inline app/scripts/lib/channels/mixins/postmessage_receiver.js]: Logged.

**Comment by shane-tomlinson** [comment]
Removed the WIP label, I'm calling this ready for review. @vladikoff and @rfk, could you give this a spin?

Testing the Sync flow is necessary since this changes the way postMessages are handled.

File: app/scripts/lib/auth-errors.js [hunk 6]
```diff
@@ -190,10 +190,6 @@ function (_, Errors) {
       errno: 1021,
       message: UNEXPECTED_ERROR_MESSAGE
     },
-    ILLEGAL_IFRAME_PARENT: {
-      errno: 1022,
-      message: t('Firefox Accounts can only be placed into an IFRAME on approved sites')
-    },
```
Comment (rfk) [inline app/scripts/lib/auth-errors.js]: Is this the only place this now-obsolete errno was documented?  If may be worth leaving a comment to explain the gap in errno numbering.

File: app/scripts/lib/channels/mixins/postmessage_receiver.js [hunk 7]
```diff
@@ -90,9 +92,37 @@ define([
...
+      // origin of the string constant 'null'. See
+      // https://developer.mozilla.org/docs/Web/API/Window/postMessage#Using_win.postMessage_in_extensions
+      // and
+      // https://bugzilla.mozilla.org/show_bug.cgi?id=1040257
+      //
+      // `message` events on Fx for iOS come from the page itself and
+      // should have the origin of the content server.
+      // See https://github.com/mozilla/firefox-ios/blob/66c8d746f0c8ca0f3c03122663205276b8ed85ab/FxAClient/Frontend/SignIn/FxAContentViewController.swift#L192
+      // `message` event from functional tests come from the page itself
+      // and should have the origin of the content server.
```
Comment (rfk) [inline app/scripts/lib/channels/mixins/postmessage_receiver.js]: It's not obvious how the "origin of the conent server" case is handled by the code below.  Since it's called out in a comment, I was expecting to see code for handling it as an expicit case.  If the overall effect of the iOS behaviour is to make everything "Just Work" then it may be worth mentioning that in the comment with e.g. "and hence does not require special handling".

File: app/scripts/lib/channels/mixins/postmessage_receiver.js [hunk 8]
```diff
@@ -90,9 +92,37 @@ define([
...
+
+      if (event.origin === 'null') {
+        return true;
+      }
```
Comment (rfk) [inline app/scripts/lib/channels/mixins/postmessage_receiver.js]: It may also be worth a more explicit comment on the logic here, to say "this can only have come from Fx Desktop, so it's safe to trust it"

File: app/tests/spec/lib/channels/fx-desktop.js [hunk 9]
```diff
@@ -92,6 +102,57 @@ function (chai, sinon, AuthErrors, FxDesktopChannel, WindowMock, TestHelpers) {
...
+          data: {
+            type: 'message',
+            content: {
+              status: 'can_link_account'
+            }
+          }
+        });
+
+        assert.isFalse(metrics.logError.called);
+        assert.isTrue(channel.trigger.called);//('can_link_account'));
```
Comment (rfk) [inline app/tests/spec/lib/channels/fx-desktop.js]: Commented out for debugging purposes?

**Comment by rfk** [comment]
Code looks good, unfortunately I haven't had a chance to try it out in a live flow and likely won't until Tuesday.

> There is no longer a special "illegal iframe" screen

Is this just to simplify and remove code, or does removing it have some other benefit?  (Don't get me wrong, I'm pro removing code!  Just curious.)

**Comment by ncalexan** [comment]
Hey all -- I just want to be clear that iOS has not shipped anything yet
and so we are completely free to make iOS do things that *don't* require
special server-side code paths.  I would make iOS implement the WebChannel
approach if we had/could get the server to do the desktop sign-up flow over
WebChannel.

File: app/scripts/lib/auth-errors.js (hunk 6 above)
Comment (shane-tomlinson) [inline app/scripts/lib/auth-errors.js]: I was wondering how to handle error removal. I could leave the error in place and change the `message` to something like `deprecated` so that it's understood to no longer be used.

**Comment by shane-tomlinson** [comment]
> > There is no longer a special "illegal iframe" screen

> Is this just to simplify and remove code, or does removing it have some other benefit? (Don't get me wrong, I'm pro removing code! Just curious.)

The overall goal is to ensure if the content server is iframed, it is iframed by a trusted origin. Since we approve all OAuth reliers, we decided if the iframe's parent is the same origin as the origin registered for the `client_id` used on startup, then we trust them.

To find the iframe parent's origin, we have to rely on something that cannot be faked. Request headers can be faked, so an alternative approach is needed. Our approach is to postMessage a `ping` out to the parent, and expect a `ping` response. A postMessage event contains `event.origin` which comes from the browser and cannot be faked.

Originally, the initial `ping` was sent to `*`. Anybody could respond, and we'd get the origin from the event. We then take the `client_id` from the URL query parameters and fetch the relier's OAuth information. If the registered origin matches the postMessage response's origin, we are good.

With this PR, we change the approach slightly. We *first* fetch the OAuth information for the `client_id`. We get the `origin` from the OAuth information and send the initial `ping` to that origin. This prevents `ping`s from going to possible bad acting origins. The relier could still send a `ping` postMessage to us, but we still check the origin and will refuse to start if the origin is not what we expect.

So that's the background. That should be put into documentation if it hasn't been already.

The `illegal_iframe` screen's intent was to notify the user if they are the target of a hack. I thought "if the relier no longer receives the `ping` message, they'll no longer respond, and we won't know the origin`. After writing all of this, I see my reasoning is faulty, the relier could still respond.

Perhaps the `illegal_iframe` screen still has use, especially for relier developers. This could be a message we expect a response to. If the response does not arrive after N seconds or if the origin does not match, redirect to the `illegal_iframe` screen and say the origin does not match the info registered for the `client_id`.

**Comment by shane-tomlinson** [comment]
> I would make iOS implement the WebChannel approach if we had/could get the server to do the desktop sign-up flow over WebChannel.

@ncalexan - I'm working on just such a thing for Fx Desktop. I'll let you know once I'm a bit further along.

**Comment by shane-tomlinson** [comment]
> After writing all of this, I see my reasoning is faulty, the relier could still respond.

You know, I knew this. And I am logging the error.

File: app/tests/spec/lib/url.js [hunk 10]
```diff
@@ -63,6 +63,15 @@ function (chai, _, Url) {
...
     });
+
+    describe('getOrigin', function () {
+      it('returns the origin portion of the URL', function () {
+        assert.equal(
+          Url.getOrigin('https://marketplace.firefox.com/redirect/to_this_page'),
+          'https://marketplace.firefox.com'
+        );
+      });
+    });
```
Comment (rfk) [inline app/tests/spec/lib/url.js]: Should we add some tests for e.g. non-standard ports and other edge-cases?

Comment (shane-tomlinson) [inline app/tests/spec/lib/url.js]: :+1:

**Comment by shane-tomlinson** [comment]
@rfk - added several additional Url.getOrigin tests, and found a bug in the process.

**Comment by rfk** [comment]
OK I am finally trying this out in a desktop sync flow.  If I change the expected string from 'null' to something else, I get the expected series messages in the debug console:
screen shot 2015-03-27 at 16 20 14

File: app/scripts/lib/channels/mixins/postmessage_receiver.js [hunk 11]
```diff
@@ -88,9 +91,35 @@ define([
...
-      // TODO - Ensure the event.origin is one we trust,
-      // esp in the iframe case
+      if (! this.isEventFromExpectedOrigin(event)) {
+        // from an unexpected origin, drop it on the ground.
+        var err = AuthErrors.toError('UNEXPECTED_POSTMESSAGE_ORIGIN');
+
+        // set the unexpected origin as the context, this will be logged.
+        err.context = event.origin;
+
+        this.window.console.error('postMessage received from unexpected origin');
```
Comment (rfk) [inline app/scripts/lib/channels/mixins/postmessage_receiver.js]: nit: we could log the actual and expected origins here as well, to assist debugging

**Comment by rfk** [comment]
With the correct 'null' string in place, desktop sign-up and sign-in flows work successfully for me.

(I get a spurious "Unauthorized" logged to the browser console on the sign-in flow, which appear to be from an attempt to load my profile picture rather than anything related to this PR).

r+

**Comment by shane-tomlinson** [comment]
> (I get a spurious "Unauthorized" logged to the browser console on the sign-in flow, which appear to be from an attempt to load my profile picture rather than anything related to this PR).

This is issue #2036.

**Comment by shane-tomlinson** [comment]
@rfk - I changed the console.error message to print both the actual and expected domains.
